### PR TITLE
Health check: check if proxies trust anchors match configuration

### LINKED ISSUE #3344: Check for incompatible trust roots
### What problem are you trying to solve?

If a user was to install linkerd, then uninstall and re-install, the old proxies would be using the old trust root. This would end up creating some weird bad certificate errors in the logs. `check` or `check --proxy` should validate that the trust roots match for every pod that has the proxy in the cluster.

### PR DESCRIPTION
**Pull request by ereslibre**
If Linkerd is reinstalled or if the trust anchors are modified while
proxies are running on the cluster, they will contain an outdated
`LINKERD2_PROXY_IDENTITY_TRUST_ANCHORS` certificate.

This changeset adds support for `linkerd check`, so it checks if there
is any proxy running on the cluster, and performing the check against
the configuration trust anchor. If there's a failure (considered non
fatal), `linkerd check` will notify the user about what pods are the
offenders (and in what namespace each one is), and also a way to
remediate the issue (restarting the pods).

Fixes #3344

Signed-off-by: Rafael Fernández López <ereslibre@ereslibre.es>

### CODE REVIEW

**Comment by grampelberg** [comment]
Awesome! This is going to help a lot of folks.

**Comment by ereslibre** [comment]
> Awesome! This is going to help a lot of folks.

:) I will check a couple of things and remove the WIP from the subject, will ping when ready. Thanks!

**Comment by ereslibre** [comment]
Website PR: https://github.com/linkerd/website/pull/545

File: pkg/healthcheck/healthcheck.go [hunk 1]
```diff
@@ -1205,6 +1214,42 @@ func (hc *HealthChecker) checkPodSecurityPolicies(shouldExist bool) error {
 	return checkResources("PodSecurityPolicies", objects, []string{fmt.Sprintf("linkerd-%s-control-plane", hc.ControlPlaneNamespace)}, shouldExist)
 }
 
+func (hc *HealthChecker) checkDataPlaneProxiesCertificate() error {
+	podList, err := hc.kubeAPI.CoreV1().Pods("").List(metav1.ListOptions{LabelSelector: k8s.ProxyDeploymentLabel})
```
Comment (adleong) [inline pkg/healthcheck/healthcheck.go]: We typically use the `ControllerNSLabel` label for finding injected pods.

File: pkg/healthcheck/healthcheck.go [hunk 2]
```diff
@@ -1205,6 +1214,42 @@ func (hc *HealthChecker) checkPodSecurityPolicies(shouldExist bool) error {
...
+		for _, containerSpec := range pod.Spec.Containers {
+			if containerSpec.Name != k8s.ProxyContainerName {
+				continue
+			}
+			for _, envVar := range containerSpec.Env {
+				if envVar.Name != identity.EnvTrustAnchors {
+					continue
+				}
+				if envVar.Value != trustAnchorsPem {
+					offendingPods = append(offendingPods, fmt.Sprintf("%s (%s)", pod.ObjectMeta.Name, pod.ObjectMeta.Namespace))
```
Comment (adleong) [inline pkg/healthcheck/healthcheck.go]: We typically format pod identifiers as `namespace/name`.  E.g.:

```
× data plane proxies are ready
    The "tracing/jaeger-7cdf9ddbb4-htspm" pod is not running
    see https://linkerd.io/checks/#l5d-data-plane-ready for hints
```

File: pkg/healthcheck/healthcheck.go (hunk 1 above)
Comment (adleong) [inline pkg/healthcheck/healthcheck.go]: we should respect the `hc.DataPlaneNamespace` flag

**Comment by ereslibre** [comment]
@adleong thanks for your review! I think I addressed all your comments.

I overlooked the data plane namespace option, I also adapted the output so if a namespace is provided we omit it from the list of offending pods. Please tell me if that's fine or if you prefer to be consistent on the output and always print the namespace along with the pod name.

Thank you!

**Comment by ihcsim** [comment]
@ereslibre This is awesome! :+1: 

When I ran some tests on your branch with deployment, statefulset and jobs, the output can appear to be a bit clutter:
```
‼ data plane proxies certificate match CA
    The following pods have old proxy certificate information; please, restart them: emoji-0, emoji-786c5d657f-l2zc4, emoji-job-qjpqr, vote-bot-7bd97dfbdc-nc4jw, vote-job-bot-dh8md, vote-sts-bot-0, voting-0, voting-6f888b8bd-x5s8c, voting-job-lbl28, web-0, web-76968bc99-8bvvm, web-job-mz28n
```

Can we use line breaks to organize the output a bit? What do you think of:
```
‼ data plane proxies certificate match CA
    The following pods have old proxy certificate information; please, restart them:
        emojivoto/emoji-0
        emojivoto/emoji-786c5d657f-l2zc4
        emojivoto/emoji-job-qjpqr
        emojivoto/vote-bot-7bd97dfbdc-nc4jw
        emojivoto/vote-job-bot-dh8md
        emojivoto/vote-sts-bot-0
        emojivoto/voting-0
        emojivoto/voting-6f888b8bd-x5s8c
        emojivoto/voting-job-lbl28
        emojivoto/web-0
        emojivoto/web-76968bc99-8bvvm
        emojivoto/web-job-mz28n
    see https://linkerd.io/checks/#l5d-data-plane-proxies-certificate-match-ca for hints
```
with:
```diff
diff --git a/pkg/healthcheck/healthcheck.go b/pkg/healthcheck/healthcheck.go
index 83891905..c991eab1 100644
--- a/pkg/healthcheck/healthcheck.go
+++ b/pkg/healthcheck/healthcheck.go
@@ -1251,7 +1251,7 @@ func (hc *HealthChecker) checkDataPlaneProxiesCertificate() error {
 	if len(offendingPods) == 0 {
 		return nil
 	}
-	return fmt.Errorf("The following pods have old proxy certificate information; please, restart them: %s", strings.Join(offendingPods, ", "))
+	return fmt.Errorf("The following pods have old proxy certificate information; please, restart them:\n\t%s", strings.Join(offendingPods, "\n\t"))
 }
 
 func checkResources(resourceName string, objects []runtime.Object, expectedNames []string, shouldExist bool) error {
```
Note that changing this output format will fail the new unit tests.

**Comment by ereslibre** [comment]
Thank you @ihcsim! I think it makes a lot of sense, updating the PR. Update to https://github.com/linkerd/website/pull/545 will follow.

**Comment by ereslibre** [comment]
Seems that tests are failing to pass because the Docker Hub is going through some technical difficulties at this time (it's returning 503's).

**Comment by ereslibre** [comment]
> please fill up this form so that we can send you some Linkerd swags.

Just did, awesome! Thank you 💯🎉

**Pull request merged**